core/peerset/src/peersstate.rs

```diff
@@ -0,0 +1,586 @@
+		}
+
+		// Note that it is possible for num_out to be strictly superior to the max, in case we were
+		// connected to reserved node then marked them as not reserved, or if the user used
+		// `force_outgoing`.
+		if *self.num_out >= self.max_out {
+			return Err(self);
+		}
+
+		Ok(self.force_outgoing())
+	}
+
+	/// Sets the peer as connected as an outgoing connection.
+	pub fn force_outgoing(self) -> ConnectedPeer<'a> {
+		let connec_state = &mut self.state.connection_state;
+		debug_assert!(!connec_state.is_connected());
+		*connec_state = ConnectionState::Out;
+
+		if !self.state.reserved {
+			*self.num_out += 1;
+		}
+
+		ConnectedPeer {
+			state: self.state,
+			peer_id: self.peer_id,
+			num_in: self.num_in,
+			num_out: self.num_out,
+			max_in: self.max_in,
+			max_out: self.max_out,
+		}
+	}
+
+	/// Tries to accept the peer as an incoming connection.
+	///
+	/// If there are enough slots available, switches the node to "connected" and returns `Ok`. If
+	/// the slots are full, the node stays "not connected" and we return `Err`.
+	///
+	/// Note that reserved nodes don't count towards the number of slots.
+	pub fn try_accept_incoming(self) -> Result<ConnectedPeer<'a>, NotConnectedPeer<'a>> {
+		if self.is_reserved() {
+			return Ok(self.force_ingoing())
+		}
+
+		// Note that it is possible for num_in to be strictly superior to the max, in case we were
+		// connected to reserved node then marked them as not reserved.
+		if *self.num_in >= self.max_in {
+			return Err(self);
+		}
+
+		Ok(self.force_ingoing())
+	}
+
+	/// Sets the peer as connected as an ingoing connection.
+	pub fn force_ingoing(self) -> ConnectedPeer<'a> {
+		let connec_state = &mut self.state.connection_state;
+		debug_assert!(!connec_state.is_connected());
+		*connec_state = ConnectionState::In;
+
+		if !self.state.reserved {
+			*self.num_in += 1;
+		}
+
+		ConnectedPeer {
+			state: self.state,
+			peer_id: self.peer_id,
+			num_in: self.num_in,
+			num_out: self.num_out,
+			max_in: self.max_in,
+			max_out: self.max_out,
+		}
+	}
+
+	/// Sets whether or not the node is reserved.
+	pub fn set_reserved(&mut self, reserved: bool) {
+		self.state.reserved = reserved;
+	}
+
+	/// Returns true if the the node is reserved.
+	pub fn is_reserved(&self) -> bool {
+		self.state.reserved
+	}
+
+	/// Returns the reputation value of the node.
+	pub fn reputation(&self) -> i32 {
+		self.state.reputation
+	}
+
+	/// Sets the reputation of the peer.
+	pub fn set_reputation(&mut self, value: i32) {
+		self.state.reputation = value;
+	}
+
+	/// Performs an arithmetic addition on the reputation score of that peer.
+	///
+	/// In case of overflow, the value will be capped.
+	/// If the peer is unknown to us, we insert it and consider that it has a reputation of 0.
+	pub fn add_reputation(&mut self, modifier: i32) {
+		let reputation = &mut self.state.reputation;
+		*reputation = reputation.saturating_add(modifier);
+	}
+}
+
+/// A peer that we have never heard of.
```

Comment:
Where does the Id come from then?

Reply:
It is given by the user. This object represents an entry not-inserted-yet, just like the `VacantEntry` object in the hash map API.

Reply:
Example usage: https://github.com/paritytech/substrate/pull/2440/files#diff-f7c3d7de5448dde5fabba2020a170033R363